src/AppInstallerCLICore/Commands/DscPackageResource.cpp

```diff
@@ -0,0 +1,565 @@
+                            {
+                                AICLI_LOG(CLI, Info, << "Reinstalling package to downgrade to desired version");
+                                data.Reinstall();
+                            }
+                        }
+                    }
+                    else
+                    {
+                        AICLI_LOG(CLI, Info, << "Installing package as it was not found");
+                        data.Install();
+                    }
+                }
+                else
+                {
+                    AICLI_LOG(CLI, Info, << "Uninstalling package as desired");
+                    data.Uninstall();
+                }
+
+                if (data.SubContext->IsTerminated())
+                {
+                    return;
+                }
+            }
+
+            WriteJsonOutputLine(context, data.Output.ToJson());
+            WriteJsonOutputLine(context, diff);
+        }
+    }
+
+    void DscPackageResource::ResourceFunctionTest(Execution::Context& context) const
+    {
+        if (auto json = GetJsonFromInput(context))
+        {
+            PackageFunctionData data{ context, json };
+
+            if (!data.Get())
+            {
+                return;
+            }
+
+            data.Output.InDesiredState(data.Test());
+
+            WriteJsonOutputLine(context, data.Output.ToJson());
+            WriteJsonOutputLine(context, data.DiffJson());
+        }
+    }
+
+    void DscPackageResource::ResourceFunctionExport(Execution::Context& context) const
+    {
+        auto json = GetJsonFromInput(context, false);
+        PackageFunctionData data{ context, json, true };
+
+        data.PrepareSubContextInputs();
+
+        if (!data.Input.UseLatest().value_or(true))
+        {
+            data.SubContext->Args.AddArg(Execution::Args::Type::IncludeVersions);
+        }
+
+        data.SubContext->Args.AddArg(Execution::Args::Type::ConfigurationExportAll);
+
+        *data.SubContext <<
+            Workflow::SearchSourceForPackageExport;
+
+        if (data.SubContext->IsTerminated())
+        {
+            context.Terminate(data.SubContext->GetTerminationHR());
+            return;
+        }
+
+        const auto& packageCollection = data.SubContext->Get<Execution::Data::PackageCollection>();
+
+        for (const auto& source : packageCollection.Sources)
+        {
+            for (const auto& package : source.Packages)
+            {
+                PackageResourceObject output;
+
+                output.Identifier(package.Id);
+                output.Source(source.Details.Name);
+
+                if (!package.VersionAndChannel.GetVersion().IsEmpty())
+                {
+                    output.Version(package.VersionAndChannel.GetVersion().ToString());
+                }
+
+                if (package.Scope != Manifest::ScopeEnum::Unknown)
+                {
+                    output.Scope(ConvertScope(Manifest::ScopeToString(package.Scope), true));

```

Comment:
If we export Scope, during apply, will it be enforced during installer selection and may cause failures? i.e. we have many installers that do not specify scope.

Reply:
Good point; I think we would need to support the "OrUnknown" variants for this to work as simply as it is now.  Alternately, we would need to check that the current scope is an available installer or track scope intent as we do for some other values.

Reply:
Yeah, the other thing (for export) besides scope that could be a challenge is the install location.